File: tasks/binary_search/binary-search.js
function search(list, elem) {
    let max = list.length - 1;
    let min = 0;

    while (min <= max) {
        const middle = Math.round(min + (max - min)/ 2);

        const value = list[middle];

        if (value < elem) {
            min = middle + 1;
        } else if (value > elem) {
            max = middle - 1;
        } else {
            return middle;
        }
    }

    return null;
}

console.log(search([1,3,5,7,9,11], 9));

File: tasks/binary_search/binary-search.erl
-module(binary_search).
-export([search/2, at/2]).

at(_, []) -> {result, null};
at(Pos, List) when Pos >= erlang:length(List) -> {result, null};
at(Pos, List) ->
    at(Pos, List, 0).
at(Pos, [H|_], Acc) when Pos =:= Acc -> {result, H};
at(Pos, [_|T], Acc) -> at(Pos, T, Acc + 1).

search([], _) -> {result, null};
search(List, Find) ->
    search(List, Find, 0, erlang:length(List)).
search(List, Find, Min, Max) ->
    Index = erlang:list_to_integer(erlang:float_to_list(Min + (Max - Min) / 2, [{decimals, 0}])),
    {result, Value} = binary_search:at(Index, List),
    if  
        Value > Find -> search(List, Find, Min, Index - 1);
        Value < Find -> search(List, Find, Index + 1, Max);
        true -> {result, Index}
    end.
